File: apps/macos_ime/Sources/VietEngine/VietTable.swift
// VietTable.swift
// ---------------
// Bảng tra: (chữ cái gốc, dấu biến âm, dấu thanh) -> ký tự Unicode tiếng Việt.
//
// Vì sao cần bảng này? Vì tiếng Việt có sẵn các ký tự dựng sẵn (precomposed) trong
// Unicode như "ế", "ữ", "ặ". Engine suy nghĩ theo (gốc + mark + tone) rồi tra ra
// ký tự cuối cùng để "gõ ra màn hình".
//
// Cách đọc bảng: với mỗi nguyên âm có dấu, ta liệt kê 6 biến thể theo 6 thanh,
// theo đúng thứ tự enum Tone: [none, acute, grave, hook, tilde, dot].

enum VietTable {

    /// Trả về ký tự tiếng Việt dựng sẵn cho tổ hợp (base + mark + tone).
    /// Nếu không phải nguyên âm hợp lệ -> trả về nil (engine sẽ giữ nguyên ký tự gốc).
    static func compose(base: Character, mark: Mark, tone: Tone) -> Character? {
        let isUpper = base.isUppercase
        let lower = Character(base.lowercased())

        guard let variants = lower.toneVariants(mark: mark) else { return nil }
        let result = variants[tone.index]
        return isUpper ? Character(result.uppercased()) : result
    }
}

private extension Tone {
    /// Vị trí của thanh trong mảng 6 biến thể.
    var index: Int {
        switch self {
        case .none:  return 0
        case .acute: return 1
        case .grave: return 2
        case .hook:  return 3
        case .tilde: return 4
        case .dot:   return 5
        }
    }
}

private extension Character {
    /// Trả mảng 6 biến thể theo thanh [ngang, sắc, huyền, hỏi, ngã, nặng]
    /// cho chữ cái gốc (đã lowercase) với dấu biến âm `mark`.
    /// Trả nil nếu tổ hợp không hợp lệ trong tiếng Việt.
    func toneVariants(mark: Mark) -> [Character]? {
        switch (self, mark) {
        // --- a ---
        case ("a", .none):       return chars("a á à ả ã ạ")
        case ("a", .circumflex): return chars("â ấ ầ ẩ ẫ ậ")
        case ("a", .breve):      return chars("ă ắ ằ ẳ ẵ ặ")
        // --- e ---
        case ("e", .none):       return chars("e é è ẻ ẽ ẹ")
        case ("e", .circumflex): return chars("ê ế ề ể ễ ệ")
        // --- i ---
        case ("i", .none):       return chars("i í ì ỉ ĩ ị")
        // --- o ---
        case ("o", .none):       return chars("o ó ò ỏ õ ọ")
        case ("o", .circumflex): return chars("ô ố ồ ổ ỗ ộ")
        case ("o", .horn):       return chars("ơ ớ ờ ở ỡ ợ")
        // --- u ---
        case ("u", .none):       return chars("u ú ù ủ ũ ụ")
        case ("u", .horn):       return chars("ư ứ ừ ử ữ ự")
        // --- y ---
        case ("y", .none):       return chars("y ý ỳ ỷ ỹ ỵ")
        // --- đ (phụ âm, không mang dấu thanh; lặp 6 lần cho khớp định dạng) ---
        case ("d", .dyet):       return chars("đ đ đ đ đ đ")
        default:
            return nil
        }
    }
}

/// Helper: tách chuỗi "á à ả" thành mảng Character.
private func chars(_ s: String) -> [Character] {
    s.split(separator: " ").map { Character(String($0)) }
}

File: packages/viet_engine/lib/src/viet_table.dart
// viet_table.dart
// ---------------
// Bảng tra: (chữ cái gốc, dấu biến âm, dấu thanh) -> ký tự Unicode tiếng Việt.
//
// Vì sao cần bảng này? Vì tiếng Việt có sẵn các ký tự dựng sẵn (precomposed) trong
// Unicode như "ế", "ữ", "ặ". Engine suy nghĩ theo (gốc + mark + tone) rồi tra ra
// ký tự cuối cùng để "gõ ra màn hình".
//
// Cách đọc bảng: với mỗi nguyên âm có dấu, ta liệt kê 6 biến thể theo 6 thanh,
// theo đúng thứ tự enum Tone: [none, acute, grave, hook, tilde, dot].

import 'viet_model.dart';

/// Mảng 6 biến thể theo thanh [ngang, sắc, huyền, hỏi, ngã, nặng]
/// cho chữ cái gốc (đã lowercase) với dấu biến âm `mark`.
/// Trả null nếu tổ hợp không hợp lệ trong tiếng Việt.
List<String>? _toneVariants(String lower, Mark mark) {
  final entry = _variants['$lower|${mark.name}'];
  return entry;
}

List<String> _chars(String s) => s.split(' ');

final Map<String, List<String>> _variants = {
  // --- a ---
  'a|none': _chars('a á à ả ã ạ'),
  'a|circumflex': _chars('â ấ ầ ẩ ẫ ậ'),
  'a|breve': _chars('ă ắ ằ ẳ ẵ ặ'),
  // --- e ---
  'e|none': _chars('e é è ẻ ẽ ẹ'),
  'e|circumflex': _chars('ê ế ề ể ễ ệ'),
  // --- i ---
  'i|none': _chars('i í ì ỉ ĩ ị'),
  // --- o ---
  'o|none': _chars('o ó ò ỏ õ ọ'),
  'o|circumflex': _chars('ô ố ồ ổ ỗ ộ'),
  'o|horn': _chars('ơ ớ ờ ở ỡ ợ'),
  // --- u ---
  'u|none': _chars('u ú ù ủ ũ ụ'),
  'u|horn': _chars('ư ứ ừ ử ữ ự'),
  // --- y ---
  'y|none': _chars('y ý ỳ ỷ ỹ ỵ'),
  // --- đ (phụ âm, không mang dấu thanh; lặp 6 lần cho khớp định dạng) ---
  'd|dyet': _chars('đ đ đ đ đ đ'),
};

/// Trả về ký tự tiếng Việt dựng sẵn cho tổ hợp (base + mark + tone).
/// Nếu không phải nguyên âm hợp lệ -> trả về null (engine sẽ giữ nguyên ký tự gốc).
String? composeViet(String base, Mark mark, Tone tone) {
  final isUpper = base.toUpperCase() == base && base.toLowerCase() != base;
  final lower = base.toLowerCase();

  final variants = _toneVariants(lower, mark);
  if (variants == null) return null;
  final result = variants[tone.index];
  return isUpper ? result.toUpperCase() : result;
}
